azavea/franklin
# Paginate search and list items responses

## Overview

This PR paginates the search and item list responses based on the ADR in #39. It's WIP because it doesn't actually do that yet, just adds types so far and tests that the opaque token generation and parsing works. It also downgrades the version of openjdk that we build in to 8, since we use that in a few other places (including in GeoTrellis CI) and don't have the same flaky memory issues that this PR has in CI.

### Checklist


### Notes

~Based on description of the [context object](https://github.com/radiantearth/stac-api-spec/blob/43799ad169da69d5558f3b492665d90bb0c63f0f/extensions/context/README.md#context-object), I think I've done the right thing with the count -- since the pagination token in our case is a filter on the results, I think it makes sense to reduce the count accordingly. Reasonable people can disagree about that though.~ (like an hour later: I no longer agree with what I said previously and I'm making the count stable regardless of what page you're on -- f4bc955)

Also, the GET search endpoint refuses to encode either the query or the `intersects` filter in the url. I think that's a sensible choice but I'm not sure how to document it. 🤔 

### Testing Instructions

- bring up your server
- browse around in the API, tossing query parameters onto `/collections/<id>/items?...` and `/search`, making sure that the next link cooperates
- create this nice json object, subbing in a reasonable collection id, in `filters.json`

```json
{
    "collections": ["a-collection-id-in-your-db"],
    "limit": 1
}
```

- `cat filters.json | http :9090/search` -- inspect the link that you get back and make sure that its extra fields make sense based on what's in the [paging extension](https://github.com/radiantearth/stac-api-spec/blob/7a6e12868113c94b17dead8989f49d978d5dd865/api-spec.md#paging-extension) description of what to return with posts
- provide some other parameters in `filters.json` to make sure that they make sense too

Closes #42
